Activity regularization

phreeza opened the pull request:
As discussed in #231 

I added a way to add updates to the global cost function from a layer. This has the advantage that we can let theano calculate the gradient, unlike the way it is in the other regularizers right now where we need to supply the gradient. Maybe all regularizers could be handled like this in the future?

I will work on an implementation of the sparse denoising autoencoder tomorrow, there is a trivial demo included already.

fchollet commented on keras/models.py:
```diff
@@ -321,6 +325,7 @@ def __init__(self):
         self.params = [] # learnable
         self.regularizers = [] # same size as params
         self.constraints = [] # same size as params
+        self.cost_updates = [] # NOT the same size as params
```
Consistency?

phreeza replied:
You mean the different lengths? I don't think it can be helped.

fchollet replied:
Yes, it's a presentation/comments issue. The comment "same size as params" means there is a 1:1 mapping between the arrays. The comment "NOT the same size as params" is not informative as to what the array contains.

fchollet commented on keras/models.py:
```diff
@@ -87,7 +87,11 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
-        updates = self.optimizer.get_updates(self.params, self.regularizers, self.constraints, train_loss)
+        if hasattr(self, 'cost_updates'):
+            for u in self.cost_updates:
+                train_loss += u()
```
Definitely a nice way to implement regularization. We'll have to think about a unified system for integrating regularization as part of the loss.

phreeza replied:
I think the existing regularisations could be turned into this style quite easily, should we do that?

fchollet replied:
Yes, that would be a good idea. We would need to monitor accuracy and speed along the way. Incorporating regularization in the loss should result in higher accuracy but *might* cause performance issues.

fchollet commented on keras/regularizers.py:
```diff
@@ -23,4 +23,23 @@ def l1l2wrap(g, p):
+    # activity dependent l1 norm
+    def l1wrap(layer):
+        # needs to be wrapped twice because input is not present during instantiation
+        def l1wrap_wrap():
```
But who will wrap the wrappers? `l1wrap_wrap`, apparently.

fchollet replied:
Would there be a better way to handle this?

phreeza replied:
Probably the way to go would be to make regularizers a class. We could have a class that includes both the old (gradient updating) and new (cost updating) functions. They would then implement either get_cost_update or get_gradient_update.

How do you feel about that? I am a bit busy right now but I could get to it around the middle of the week.

phreeza commented:
I have refactored the regularizer system to be object oriented, I think it is more elegant and futureproof this way. I managed to keep it backwards compatible, so these changes should not break any existing code.

It leaves the option open to transition to purely cost-based regularization in the future, if desired.

I also added proper unit tests. These use the old style of creating regularizers.

fchollet commented on keras/regularizers.py:
```diff
@@ -1,26 +1,117 @@
+        raise NotImplementedError
+
+
+class RegularizerWeightsL1(Regularizer):
```
What about simply `L1`? It lives inside `regularizers` and inherits from `Regularizer`, so it's pretty clear it's a regularizer.

fchollet commented on keras/layers/containers.py:
```diff
@@ -20,6 +20,8 @@ def __init__(self, layers=[]):
         self.params = []
         self.regularizers = []
         self.constraints = []
+        self.cost_updates = []
```
The general term used throughout Keras is "loss" (as well as "objective" for the objective functions); I think we should keep it that way and replace "cost" with "loss" in the variables.

fchollet commented on keras/models.py:
```diff
@@ -89,7 +89,12 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
+        if hasattr(self, 'cost_updates'):
+            for u in self.cost_updates:
+                if u.modifies_cost:
+                    train_loss = u(train_loss)
```
You can get rid of this parameter (and its check in `compile`) by using:
`train_loss += u.loss() `

(which would return 0 if the update does not affect the loss).

fchollet commented on keras/regularizers.py:
```diff
@@ -1,26 +1,117 @@
+l1 = RegularizerWeightsL1
+l2 = RegularizerWeightsL2
+l1l2 = RegularizerWeightsL1L2
+identity = RegularizerIdentity()
```
That's cool. For consistency with activity we should have something like

`l2 = weights_l2 = WeightsL2`

fchollet commented:
I like the move to a `Regularizer` class. I left a few comments!

phreeza commented:
Hi François, thanks for your feedback.

I have fixed the other suggestions, but I am a bit unsure what you mean about removing the flags. The advantage of `train_loss = u(train_loss)` instead of  `train_loss += u.loss()` is that it allows more complex interactions such as setting the gradient to zero. I am sure there are other scenarios where something more than additive cost functions might be useful?

But of course those two flags `modifies_cost` and `modifies_gradient` are needed to achieve that. I feel like they can also serve as an additional sanity check. I was also thinking the flags could make it possible to store all cost updates in a single list at some point in the future.

Let me know what you think, Thomas

fchollet commented:
> The advantage of train_loss = u(train_loss) instead of train_loss += u.loss() is that it allows more complex interactions such as setting the gradient to zero. I am sure there are other scenarios where something more than additive cost functions might be useful?

Right, in that case a better syntax would be `train_loss = u.update_loss(train_loss)`. 

What I mean by getting of the flags is:
```python
# instead of:
if u.modifies_loss:
   loss = u(loss)

# use
loss = u.update_loss(loss) # can also leave the loss unchanged...

# instead of:
if u.modifies_gradient:
   g, p = u(g, p)

# use
g, p = u.update_gradient(g, p) # can also leave g, p unchanged...
```

If does the same, but without any control flow necessary.

phreeza commented:
OK, now I understood and fixed that as well. The only downside I see with this is that it breaks any custom regularisers that people may have written earlier. But I agree this makes the structure clearer when using them.

fchollet commented on keras/regularizers.py:
```diff
@@ -1,26 +1,80 @@
+l1 = weights_l1 = WeightsL1
+l2 = weights_l2 = WeightsL2
+l1l2 = weights_l1l2 = WeightsL1L2
+identity = Identity()
```
Should be the class `Identity`, not an instance thereof...?

phreeza replied:
This is for backwards compatibility, because the old identity was not a wrapped function, while the others were. I am fine with breaking that if you are, would be more consistent.

fchollet replied:
Let's definitely prioritize consistency and code quality over backwards compatibility. Keras is in alpha and subject to occasional breaking changes.

fchollet commented on keras/regularizers.py:
```diff
@@ -1,26 +1,80 @@
+        return gradient
+
+
+class Identity(Regularizer):
```
Could definitely also implement `update_loss` (return loss unchanged).

phreeza replied:
Indeed, actually could do that in the base class.

But do we not maybe want to keep it like this so that an error gets raised in the event that one of these regularizers gets put into a place where it does not belong? (like putting an activity regularizer in the slot for a weight regularizer?)

fchollet replied:
The idea would be that Identity could serve as both the identity weight regularizer and the identity activity regularizer.

fchollet commented on keras/models.py:
```diff
@@ -370,4 +379,33 @@ def load_weights(self, filepath):
+        f.close()
+
+
+class Autoencoder(Sequential):
```
What is this model doing in this PR, given that autoencoders are being implemented with a layer in `core.py`...?

phreeza replied:
Ugh, something must have gone wrong when merging recent changes from master...

fchollet commented on keras/models.py:
```diff
@@ -89,7 +93,11 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
         self.class_mode = class_mode
 
-        updates = self.optimizer.get_updates(self.params, self.regularizers, self.constraints, train_loss)
+        if hasattr(self, 'cost_updates'):
```
Should be `loss_updates`

fchollet commented on keras/layers/core.py:
```diff
@@ -80,8 +80,6 @@ def __init__(self, models, mode='sum'):
             mode: {'sum', 'concat'}
         '''
-        if len(models) < 2:
-            raise Exception("Please specify two or more input models to merge")
```
Why remove this check?

phreeza replied:
Sorry, I accidentally included my old autoencoder design in this PR, this was part of it. No idea how that happened, should be all fixed now.

fchollet commented on tests/auto/test_regularizers.py:
```diff
@@ -0,0 +1,62 @@
+max_train_samples = 5000
+max_test_samples = 1000
+
+np.random.seed(1337) # for reproducibility
```
This line actually belong before any imports from layers.core.

fchollet commented:
Looks good! We'll merge this. Apart from the small details we mentioned, we should definitely update the doc to introduce the changes.

fchollet commented on keras/models.py:
```diff
@@ -7,7 +7,9 @@
 
 from . import optimizers
 from . import objectives
-from . import callbacks as cbks
```
This line is required...

fchollet commented:
I remove the loss_update system, which did not make sense to me in its current form. Which should reintroduce activity regularization as part of a unified way to handle regularization via the loss function. I'll look at it later today.

fchollet commented:
Here's what I propose: regularizers get instantiated with the quantity that they regularize (weight parameter or layer output). A list of such regularizers (of arbitrary length) is then used to update the loss function of a model.
```python
class Regularizer(object):
    def __call__(self, loss):
        return loss


class WeightRegularizer(Regularizer):
    def __init__(self, p, l1=0., l2=0.):
        self.p = p
        self.l1 = l1
        self.l2 = l2

    def __call__(self, loss):
        loss += T.sgn(self.p) * self.l1
        loss += self.p * self.l2
        return loss


class ActivityRegularizer(Regularizer):
    def __init__(self, output, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2
        self.output = output

    def set_layer(self, layer):
        self.layer = layer

    def __call__(self, loss):
        loss += self.l1 * T.sum(T.mean(abs(self.output), axis=0))
        loss += self.l2 * T.sum(T.mean(self.output ** 2, axis=0))
        return loss
```

fchollet commented:
Done, all regularization is now happening via the loss function. Activity regularization is back as well, both through the dedicated ActivityRegularization layer, and through arguments allowed in certain layers (Dense layers, Convolution layers).

kenterao commented:
I agree that making regularizer classes makes things cleaner.  

What do you think about adding ```get_config() ``` to the ```Regularizer``` class so a layer can be fully reconstructed from it's config.  

```python
class WeightRegularizer(Regularizer):
    def __init__(self, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2

    def get_config(self):
        return {'name': self.__class__.__name__,
            'l1': self.l1, 'l2': self.l2}

class Dense(Layer):
    def get_config(self):
        return {"name": self.__class__.__name__,
            "W_regularizer": self.W_regularizer.get_config(),
            "b_regularizer": self.b_regularizer.get_config(),
            ...
        }
```

phreeza commented:
I like what you did with the regularizers, @fchollet , I guess I was too worried about compatibility, keeping the gradient-updating regularizers :-)

@kenterao's suggestion sounds good to me.
